core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -540,4 +556,102 @@ class EntityService(protected val ctx: RawlsRequestContext,
       )
   }
 
+  /**
+    * Migrate all entity data in a given workspace from legacy (LocalEntityProvider) to compact (Quicksilver) format.
+    *
+    * This migration method is unoptimized and in flux; use at your own risk
+    *
+    */
+  def quicksilverMigration(workspaceName: WorkspaceName): Future[Map[String, Int]] = {
+    implicit val system: ActorSystem = ActorSystem("quicksilverMigration")
+
+    for {
+      // verify owner of workspace.
+      // TODO CORE-364: require some kind of admin permission via asFCAdmin or a resource type admin instead?
+      workspaceContext <- getV2WorkspaceContextAndPermissions(workspaceName,
+                                                              SamWorkspaceActions.own,
+                                                              Some(WorkspaceAttributeSpecs(all = false))
+      )
+
+      // confirm if this is already a quicksilver workspace by checking settings
+      workspaceSettingService = workspaceSettingServiceConstructor.get.apply(ctx)
+      settings <- workspaceSettingService.getWorkspaceSettings(workspaceName)
+      _ = if (
+        settings
+          .find(_.isInstanceOf[CompactDataTablesSetting])
+          .asInstanceOf[Option[CompactDataTablesSetting]]
+          .exists(_.config.enabled)
+      ) {
+        throw new RawlsExceptionWithErrorReport(ErrorReport("Quicksilver already enabled for this workspace"))
+      }
+
+      // get the local (legacy) provider
+      localProvider <- entityManager.resolveProviderFuture(EntityRequestArguments(workspaceContext, ctx))
+
+      // change the workspace to be quicksilver-enabled
+      _ <- workspaceSettingService.setWorkspaceSettings(
+        workspaceName,
+        List(CompactDataTablesSetting(CompactDataTablesConfig(enabled = true)))
+      )
```

Comment:
Could we convert the workspace to `quicksilver-enabled` after a successful migration in case of a failure in the migration transaction?

Reply:
good idea, changed